rust/worker/src/execution/operators/brute_force_metadata_filtering.rs

```diff
@@ -0,0 +1,718 @@
+                        record_segment_reader = None;
+                    }
+                    RecordSegmentReaderCreationError::BlockfileOpenError(e) => {
+                        tracing::error!("Error creating record segment reader {}", e);
+                        return Err(BruteForceMetadataFilteringError::BruteForceMetadataRecordSegmentReaderCreationError(
+                            RecordSegmentReaderCreationError::BlockfileOpenError(e),
+                        ));
+                    }
+                    RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
+                        tracing::error!("Error creating record segment reader {}", e);
+                        return Err(BruteForceMetadataFilteringError::BruteForceMetadataRecordSegmentReaderCreationError(
+                            RecordSegmentReaderCreationError::InvalidNumberOfFiles,
+                        ));
+                    }
+                };
+            }
+        };
+        let materializer = LogMaterializer::new(
+            record_segment_reader,
+            input.log_record.clone(),
+            input.curr_max_offset_id.clone(),
+        );
+        let mat_records = match materializer.materialize().await {
+            Ok(records) => records,
+            Err(e) => {
+                return Err(
+                    BruteForceMetadataFilteringError::BruteForceMetadataLogMaterializationError(e),
+                );
+            }
+        };
+        let mut ids_to_metadata: HashMap<u32, HashMap<&String, &MetadataValue>> = HashMap::new();
+        for (records, _) in mat_records.iter() {
+            if records.final_operation == Operation::Delete {
+                continue;
+            }
+            if !ids_to_metadata.contains_key(&records.offset_id) {
+                ids_to_metadata.insert(records.offset_id, HashMap::new());
+            }
+            let map_pointer = ids_to_metadata.get_mut(&records.offset_id).expect(
+                "Just inserted the key one line above so cannot happen that it does not exist now",
+            );
+            match &records.data_record {
+                Some(data_record) => match &data_record.metadata {
+                    Some(meta) => {
+                        for (meta_key, meta_val) in meta {
+                            map_pointer.insert(&meta_key, &meta_val);
+                        }
+                    }
+                    None => (),
+                },
+                None => (),
+            };
+            match &records.metadata_to_be_merged {
+                Some(meta) => {
+                    for (meta_key, meta_val) in meta {
+                        map_pointer.insert(meta_key, meta_val);
+                    }
+                }
+                None => (),
+            };
+        }
+        // Time now to perform a metadata search based on where clause.
+        let clo = |metadata_key: &str,
```

Comment:
Nit: Is this the exact same callback as elsewhere? If so we could probably pull it out and reuse it.

Reply:
No this is different. We only invoke `process_where_document_with_callback` twice - once when querying from the metadata segment and the other time when querying from brute force logs. And though the structure looks similar, both the callbacks are different because one needs to read from storage and other needs to read from in-memory maps created during brute forcing.